### DISCUSSION TITLE
[Feature-Request] Request tracing support for Kong's log pdk

### DISCUSSION THREAD

**Discussion by rishabh-gupta2** [Kong/kong#7347] [Ideas and feature requests]
### Summary
Kong provides functionality for distributed request tracing with the use of plugin `correlation-id`.
While it supports sending headers to upstream and downstream, Kong's log pdk does not make use of the correlation-id. Would it make sense to have request tracing functionality within Kong's logging pdk?

If the plugin is enabled, the logging pdk can use the trace id set in shared ctx. If not enabled, the pdk can function the same way. This would definitely require some changes in the plugin itself like changing the priority order (moving it up the execution order) and setting its trace id as part of shared request ctx.

What are your thoughts on this?

**Comment by bungle** [comment]
@rishabh-gupta2, AFAIK we added extensible log serializer, so I guess best option would be to make `correlation-id` plugin to add that field to serializer?

**Comment by rishabh-gupta2** [comment]
@bungle : The requirement is to add request tracing support to usages such as `kong.log.info('sample', 'log')`, `kong.log.err('error', 'log')`. Extending the log serializer won't really work for us as it is only valid for usages of `kong.log.serialize()`.  Moreover `kong.log.serialize()` I think is a heavier operation as compared to `kong.log.info`, `kong.log.err`

**Comment by kikito** [comment]
@rishabh-gupta2 Kong currently supports changing the log format through Nginx variables (see the `log_format` option described here: https://docs.konghq.com/gateway-oss/2.4.x/logging/#removing-certain-elements-from-your-kong-logs).

If I understand correctly what you would need is an extra nginx variable set by the correlation_id plugin so that it can then be included in the `log_format` option. Is that correct?

**Reply by timorantalaiho** [reply]
I am not the original feature request author, but would answer yes :) For us, making it possible to add correlation id variable to nginx `log_format` would be clearly beneficial.

**Comment by ealogar** [comment]
Afaik, a custom log_format is only available at access_log and the original authors, requests for tracing support in error_log:
http://nginx.org/en/docs/ngx_core_module.html#error_log
vs
http://nginx.org/en/docs/http/ngx_http_log_module.html#access_log
